Storage/Logins.swift

```diff
@@ -0,0 +1,239 @@
+            let url1 = NSURL(string: self.formSubmitUrl!)
+            if url1?.host != url2?.host {
+                assertionFailure("Form submit url domain doesn't match login's domain")
+                formSubmitUrl = nil
+            }
+        }
+    }
+
+    // SyncableLoginData
+    public var guid: String? = nil
+    public var isDeleted: Bool = false
+
+    // UsageLignData
+    public var timeCreated = NSDate.nowMicroseconds()
+    public var timeLastUsed = NSDate.nowMicroseconds()
+    public var timePasswordChanged = NSDate.nowMicroseconds()
+
+    // Printable
+    public var description: String {
+        return "Login: \(username) for \(hostname)"
+    }
+
+    public class func createWith(hostname: String, username: String, password: String) -> LoginData {
+        return Login(hostname: hostname, username: username, password: password) as LoginData
+    }
+
+    public class func createWith(credential: NSURLCredential, protectionSpace: NSURLProtectionSpace) -> LoginData {
+        return Login(credential: credential, protectionSpace: protectionSpace) as LoginData
+    }
+
+    init(hostname: String, username: String, password: String) {
+        self.credentials = NSURLCredential(user: username, password: password, persistence: NSURLCredentialPersistence.None)
+        self.protectionSpace = NSURLProtectionSpace(host: hostname, port: 0, `protocol`: nil, realm: nil, authenticationMethod: nil)
+    }
+
+    init(credential: NSURLCredential, protectionSpace: NSURLProtectionSpace) {
+        self.credentials = credential
+        self.protectionSpace = protectionSpace
+    }
+
+    public func toDict() -> [String: String] {
+        return [
+            "hostname": hostname,
+            "formSubmitURL": formSubmitUrl ?? "",
+            "httpRealm": httpRealm ?? "",
+            "username": username ?? "",
+            "password": password,
+            "usernameField": usernameField ?? "",
+            "passwordField": passwordField ?? ""
+        ]
+    }
+
+    public class func fromScript(url: NSURL, script: [String: String]) -> LoginData {
+        let login = Login(hostname: getPasswordOrigin(url.absoluteString!)!, username: script["username"]!, password: script["password"]!)
+
+        if let formSubmit = script["formSubmitUrl"] {
+            login.formSubmitUrl = formSubmit
+        }
+
+        if let passwordField = script["passwordField"] {
+            login.passwordField = passwordField
+        }
+
+        if let userField = script["usernameField"] {
+            login.usernameField = userField
+        }
+
+        return login as LoginData
+    }
+
+    private class func getPasswordOrigin(uriString: String, allowJS: Bool = false) -> String? {
+        var realm: String? = nil
+        if let uri = NSURL(string: uriString) {
+            if allowJS && uri.scheme == "javascript" {
+                return "javascript:"
+            }
+
+            realm = "\(uri.scheme!)://\(uri.host!)"
```

Comment:
I don't think so. We pass in the original (toplevel) page here. I'm not sure when its scheme or host could ever need escaping...